### ISSUE TITLE
Semi-transparent-cover doesn't cover everything and renders wrongly

### ISSUE THREAD

**Issue by JoshuaLampert** [touying-typ/touying#112]
You can find an MWE at : https://typst.app/project/rO30herYanvASbIkJdjjW2.
This issue has two (maybe related?) sub-issues:
1. In the first subslide we see that the lower line of the equation is not central, the "x" is not in math font, and the vector is smaller than in should be.
2. Not everything is covered. Only a strip, which is not high enough covers the central part of the equation, but the upper and lower parts are still visible without transparency.

**Comment by OrangeX4** [comment]
This is due to the fact that currently semi-transparent cover is very hacky and may be difficult to fix. Maybe we can consider changing the font color instead of using a semi-rect

**Comment by zral0kh** [comment]
well it should be fixed soon with #328 now being merged.

that was quite the effort and its still not perfect due to typst internal stuff, but it should suffice for most use cases.
The `color-changing-cover` and `alpha-changing-cover` can now also revert to that as fallbacks for covering blocks properly.

Note: I did somehow think we could not extract the alignment and reconstruct those things like figure, place, etc. but ofc we can. 
If you would want to properly do that we should also go into those functions and only cover the individual elements inside.
But then to use semi-rect as fallback for the text-color changing covers we also need to track that...
I will put that onto my todo list. but will likely take a while.

**Comment by JoshuaLampert** [comment]
What is the status of this issue? I updated the MWE from above to use the latest touying version and as you can see, now the equations are not covered at all. Is that expected or do I need to change anything, @zral0kh?

**Comment by zral0kh** [comment]
Yes that is expected as of rn. i will have to do some more stuff to properly cover equations since sth like vectors don't expand the line height etc. and even figures are not covered were nicely right now as the entire block is covered with one rect, not the image once and the text extra.

i am planning to fix this this week bc i cannot do much in the coming month

**Comment by zral0kh** [comment]
so i fiddled a bit around and i have come to the conclusion that either we need to restructure quite a lot and pass sth through that we don't really want, or we deprecate cover with rect and only really use it for graphics elements and favuor the `alpha-changing-cover`. we cannot reliably do equations or list markers, or rather it's not easily achievable. with alpha-changing-cover it is, since the markers are text and equations are rendered internally via text.

i updated the `alpha-changing-cover` so it now properly works for equations. and i will make it also apply alpha to other colors like fill, perhaps i will just do a fallback-hide `auto` which is the semi-transparent cover which then will only get called on content that we cannot affect with alpha at all: graphics of all kinds. this should then finally give us a proper alpha cover. 

the color changing cover will then do a similar thing and also transform the other colors by the given color. behaviour would be that we use the color difference between the given color and the current text color at that point to infer the color shift. (computed in oklab)

**Comment by JoshuaLampert** [comment]
Thanks for looking into it! I'm happy to use `alpha-changing-cover` instead of `semi-transparent-cover`. In fact in touying v0.6.1 it did exactly what I expect, but starting from v0.6.2 the parentheses of the vector are rendered wrongly (see the current version of the MWE above). So fixing this to render as in v0.6.1 would already be fine with me.

**Comment by zral0kh** [comment]
please test the implementation

**Comment by JoshuaLampert** [comment]
Yes, the version from #370 looks fine to me using `alpha-changing-cover`. Thanks!

**Comment by JoshuaLampert** [comment]
Out of curiosity: When would `semi-transparent-cover` be preferred over `alpha-changing-cover`?

**Comment by zral0kh** [comment]
idk basically never i guess? it is now meant for covering context results, images and other block level content that has no accessible colors via typst.

perhaps if you have another package that relies on context and parses its stuff before touying then we fallback to that one.
i am pretty sure you would no longer want to use it for everything, keeping it as the fallback is enough.

as you can see having lists of varying type directly after each other is currently a bit broken, they get sometimes joined together on list type boundaries, not sure where this happens tho. but perhaps we should ship this improvement before this bug is fixed? @OrangeX4 what do you think?

**Comment by OrangeX4** [comment]
I think it's okay. We can fix it one by one.

**Comment by zral0kh** [comment]
gonna close this now as it has been superseeded by the solution via `alpha-changing-cover` and `color-changing-cover`.
